3.13/engine.md

```diff
@@ -0,0 +1,270 @@
+we will probably change the executable format.
+For the JIT, we will use copy-and-patch to convert the micro-ops to machine code.
+
+### Exits from executors
+
+On creation of an executor, there will be a number of potential exits
+from that executor. For each entry to an executor, exactly one exit
+will occur. This means that a few of the exits will be hot, but most will be cold.
+
+We want hot exits to be implemented such that transfer to the next executor is fast.
+However, we want cold exits to consume as little memory as possible.
+Unfortunately we cannot know in advance which are which, although there are
+some exits we expect to be very rare. We can handle these very rare exits
+differently, to save space.
+
+### Linking executors
+
+When an exit from an executor gets hot we need to create a new executor
+to attach to that exit. Once we have done that we need to link the exit to 
+the new executor in a way that allows fast transfer of execution.
+
+### Making progress
+
+If an executor exits before it does any work *and* it exits to another
+executor that can also exit before it does any work, or it exits to itself,
+we can find ourselves in an infinite loop.
+
+There are two possible approaches to avoiding this:
+1. We track which executors might not make
+progress and are careful about which executors we linked together, or
+2. We require all executors to make progress.
+
+1 has many edge cases which makes it hard to reason about and get correct.
+
+2 is simpler, but may be a bit slower as we may need to pessimize the first instruction.
+
+We will use approach 2. It is much easier to reason about and should be almost as fast.
+
+If all executors are guaranteed to make progress, then transfers from one executor 
+to another can be implemented by a single jump/tail-call.
+
+### Inter-superblock optimization
+
+Some superblocks can be quite short, but form a larger region of hot code, that 
+we want to optimize. In order to do that we want to propogate type information
+and representation changes across edges, which means storing that information
+on exits when they are cold. Since most exits will remain cold, we need to 
+store this information in a compact form.
+
+### Making "hot" exits fast and "cold" exits small
+
+In order to make hot exits fast, we them to be as simple as possible,
+passing as little information as possible.
+
+We also want to make cold exits as small as possible, but cold exits
+may become hot exits, so we need them both to use the same interface.
+
+## The implementation
+
+This section describes one possible implemnentation. The initial implementation
+will not support inter-superblock optimization, but we should plan to support it
+in the future.
+
+### Making "hot" exits fast
+
+In order to make hot exits fast, we want to implement them as a single, unconditional jump.
+In the tier 2 interpreter, that will be implemented by setting the IP to the first
+instruction of the target executor.
```

Comment:
Isn't this more specifically also decrefing the old executor, increfing the new executor etc as described below?  (And I think also updating a "current executor" pointer, correct?)

Maybe it's not relevant to repeat here, but generally speaking, it's "moving the IP + some bookkeeping".  Or is the bookkeeping not necessary when hot for some non-obvious reason?